### PR TITLE
feat: allow multiple handlers for single `try`-block

### PR DESCRIPTION
**Pull request by jaschdoc**
Closes https://github.com/flix/flix/issues/7079

### CODE REVIEW

File: main/src/ca/uwaterloo/flix/language/phase/Parser.scala [hunk 1]
```diff
@@ -1027,17 +1027,27 @@ class Parser(val source: Source) extends org.parboiled2.Parser {
...
 
-      rule {
+      def NormalTry: Rule1[ParsedAst.Expression.Try] = rule {
         SP ~ keyword("try") ~ WS ~ Expression ~ optWS ~ Body ~ SP ~> ParsedAst.Expression.Try
       }
+
+      def ChainedTry: Rule1[ParsedAst.Expression.TryChainedHandlers] = rule {
+        SP ~ keyword("try") ~ WS ~ Expression ~ optWS ~ HandlerBody ~ optWS ~ oneOrMore(HandlerBody).separatedBy(optWS) ~ SP ~>
+          ((sp1: SourcePosition, exp: ParsedAst.Expression, handler1: ParsedAst.CatchOrHandler.Handler, otherHandlers: Seq[ParsedAst.CatchOrHandler.Handler], sp2: SourcePosition) =>
+            ParsedAst.Expression.TryChainedHandlers(sp1, exp, Seq.concat(Seq(handler1), otherHandlers), sp2))
```
Comment (jaschdoc) [inline main/src/ca/uwaterloo/flix/language/phase/Parser.scala:1045]: Maybe this could be simplified by just adding `handler1` to the case class

**Comment by mlutze** [comment]
Why do we need separate chained and unchained tries?

**Comment by jaschdoc** [comment]
> Why do we need separate chained and unchained tries?

Do you want to desugar during parsing or what do you have in mind? :)

**Comment by jaschdoc** [comment]
> > > Why do we need separate chained and unchained tries?
> > 
> > 
> > Do you want to desugar during parsing or what do you have in mind? :)
> 
> It seems like `Try` with only one `With` is just a special case of `Try` with `n` `With`s.

True, but I think @magnus-madsen  just wanted to desugar it and have simpler typing rules? Could this partially solve the problem with the substitution we talked about last Friday?

**Comment by mlutze** [comment]
> > > > Why do we need separate chained and unchained tries?
> > > 
> > > 
> > > Do you want to desugar during parsing or what do you have in mind? :)
> > 
> > 
> > It seems like `Try` with only one `With` is just a special case of `Try` with `n` `With`s.
> 
> True, but I think @magnus-madsen just wanted to desugar it and have simpler typing rules? Could this partially solve the problem with the substitution we talked about last Friday?

Desugaring is fine. I just think that we don't need the "unchained" version in the parser.

**Comment by jaschdoc** [comment]
Ah okay

**Comment by jaschdoc** [comment]
Should we allow combining `catch` and `with`?
```scala
try { ... } catch { ... } with { ... }
```

**Comment by mlutze** [comment]
> Should we allow combining `catch` and `with`?
> ```scala
> try { ... } catch { ... } with { ... }
> ```

Probably not :D

**Comment by jaschdoc** [comment]
Ok I'm currently simplifying the parser but it seems weeding just becomes much more complicated. Lots of edge cases to handle the rules from the main branch. Is that worth it?

**Comment by jaschdoc** [comment]
RFR

**Comment by jaschdoc** [comment]
Weeding becomes significantly more complicated if we collapse the parsing into one rule

**Comment by magnus-madsen** [comment]
Can we not parse: `Try ... (CatchHandlerList | WithHandlerList)`?

**Comment by jaschdoc** [comment]
> Is it because of some idea that we must desugar later?

Yes but also because we handle all parsing with 1 rules so everything has to be enforced during weeding

**Comment by magnus-madsen** [comment]
What if we add ParsedAst.CatchList and ParsedAst.HandlerList?

**Comment by magnus-madsen** [comment]
A try block can have multiple catch blocks with the exact same desugaring 👍

**Comment by magnus-madsen** [comment]
Those two tricks should help alot?

**Comment by jaschdoc** [comment]
> Can we not parse: `Try ... (CatchHandlerList | WithHandlerList)`?

That could improve the solution. There is also a much simpler solution in 125d217

**Comment by jaschdoc** [comment]
I will try the `CatchHandlerList | WithHandlerList` solution now :)

**Comment by magnus-madsen** [comment]
@mlutze Ill let you take a stab and then I can come back.

File: main/src/ca/uwaterloo/flix/language/phase/Desugar.scala [hunk 2]
```diff
@@ -689,10 +689,13 @@ object Desugar {
...
 
-    case WeededAst.Expr.TryWith(exp, eff, rules, loc) =>
+    case WeededAst.Expr.TryWith(exp, handlers, loc) =>
       val e = visitExp(exp)
-      val rs = rules.map(visitHandlerRule)
-      Expr.TryWith(e, eff, rs, loc)
+      handlers.foldLeft(e) {
+        case (acc, handler) =>
+          val rs = handler.rules.map(visitHandlerRule)
+          Expr.TryWith(acc, handler.eff, rs, loc.asSynthetic)
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/Desugar.scala:697]: I think this can be a "real" location.

**Review by mlutze — APPROVED** [review]

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 3]
```diff
@@ -1040,12 +1040,12 @@ object ParsedAst {
     /**
       * Try Expression.
       *
-      * @param sp1            the position of the first character in the expression.
-      * @param exp            the guarded expression.
-      * @param catchOrHandler the handler (catch/with) of the try expression.
-      * @param sp2            the position of the last character in the expression.
+      * @param sp1      the position of the first character in the expression.
+      * @param exp      the guarded expression.
+      * @param handlers list of the handlers (catch/with) of the try expression.
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1045]: the list of catch or with blocks of the try expression.

File: main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala [hunk 4]
```diff
@@ -1873,26 +1873,47 @@ object ParsedAst {
...
+    case class Catch(rules: Seq[ParsedAst.CatchRule])
 
     /**
       * A `with` block for handling Flix effects.
       *
       * @param eff   the effect to be handled.
       * @param rules the handler rules.
       */
-    case class Handler(eff: Name.QName, rules: Option[Seq[ParsedAst.HandlerRule]]) extends CatchOrHandler
+    case class WithHandler(eff: Name.QName, rules: Option[Seq[ParsedAst.HandlerRule]])
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1915]: Why are rules here optional?

Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1915]: Not sure, it was like this before. Maybe because you can have an empty handler?

Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1915]: IMO don't complicate the parser to make types nice in ParsedAst.

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1915]: But why is it optional in the parser?

Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1915]: Because you can do `with E` (no braces) when `E` has no ops

Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1915]: I guess it's a design decision. Is an empty effect like an empty enum?

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala:1915]: I would drop it. We can reintroduce if the need arise.

With empty enums there was actually a use case.

Comment (jaschdoc) [reply main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala]: Apparently we allowed this
```scala
// Test.Exp.Effect.flix
    // @test
    pub def handleEff03(): Bool = {
        try true with Empty
    }
```

Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/ParsedAst.scala]: Yeah, I don't think we should do that :)

**Comment by magnus-madsen** [comment]
Test failires

**Comment by jaschdoc** [comment]
Yes this is unsolvable right now as discussed earlier (type inference problem) so the feature is here but does not work :/

**Comment by magnus-madsen** [comment]
> Yes this is unsolvable right now as discussed earlier (type inference problem) so the feature is here but does not work :/

Change the tests? Catch the same effect?

File: main/test/flix/Test.Exp.TryWith.flix [hunk 5]
```diff
@@ -90,23 +90,114 @@ mod Test.Exp.TryWith {
...
+        try {
+            greetAll("Cheshire Cat" :: "Queen of Hearts" :: "White Rabbit" :: "Dormouse" :: Nil)
+        } with Greet {
+            def greet(_, _cont) = ()
+        } with Greet {
+            def greet(_, _cont) = ()
+        }
+    }
+
+    // TODO disabled until better type system is in
```
Comment (magnus-madsen) [inline main/test/flix/Test.Exp.TryWith.flix:182]: I prefer not to have TODOs and comment code. Just delete what does not work or adjust it :)

Comment (magnus-madsen) [reply main/test/flix/Test.Exp.TryWith.flix:182]: It also messes with our linecount etc.

Comment (jaschdoc) [reply main/test/flix/Test.Exp.TryWith.flix:182]: Roger

**Review by magnus-madsen — APPROVED** [review]

**Comment by magnus-madsen** [comment]
Thanks!

**Pull request merged**